lib/ddtrace/contrib/rails/core_extensions.rb

```diff
@@ -1,50 +1,124 @@
+        patch_partial_renderer(::ActionView::Partials::PartialRenderer)
+      else
+        Datadog::Tracer.log.debug('Expected Template/Partial classes not found; template rendering disabled')
+      end
     end
 
-    def patch_renderer_render_template
-      if defined?(::ActionView::Renderer)
-        ::ActionView::Renderer.class_eval do
-          alias_method :render_template_without_datadog, :render_template
-          def render_template(*args, &block)
-            ActiveSupport::Notifications.instrument('start_render_template.action_view')
-            render_template_without_datadog(*args, &block)
+    def patch_template_renderer(klass)
+      klass.class_eval do
+        def render_with_datadog(*args, &block)
+          # create a tracing context and start the rendering span
+          # NOTE: Rails < 3.1 compatibility: preserve the tracing
+          # context when a partial is rendered
+          @tracing_context ||= {}
+          if @tracing_context.empty?
+            ::ActiveSupport::Notifications.instrument(
+              '!datadog.start_render_template.action_view',
+              tracing_context: @tracing_context
+            )
           end
+          render_without_datadog(*args)
+        rescue Exception => e
+          # attach the exception to the tracing context if any
+          @tracing_context[:exception] = e
+          raise e
+        ensure
+          # ensure that the template `Span` is finished even during exceptions
+          ::ActiveSupport::Notifications.instrument(
+            '!datadog.finish_render_template.action_view',
+            tracing_context: @tracing_context
+          )
         end
-      else # Rails < 3.1
-        ::ActionView::Template.class_eval do
-          alias_method :render_template_without_datadog, :render
-          def render(*args, &block)
-            ActiveSupport::Notifications.instrument('start_render_template.action_view')
-            render_template_without_datadog(*args, &block)
-          end
+
+        def render_template_with_datadog(*args)
+          # arguments based on render_template signature (stable since Rails 3.2)
+          template = args[0]
+          layout_name = args[1]
+
+          # update the tracing context with computed values before the rendering
+          template_name = template.try('identifier')
+          template_name = Datadog::Contrib::Rails::Utils.normalize_template_name(template_name)
+          layout = if layout_name.is_a?(String)
+                     # NOTE: Rails < 3.1 compatibility: the second argument is the layout name
+                     layout_name
+                   else
+                     layout_name.try(:[], 'virtual_path')
+                   end
+          @tracing_context[:template_name] = template_name
+          @tracing_context[:layout] = layout
+        rescue StandardError => e
+          Datadog::Tracer.log.error(e.message)
+        ensure
+          render_template_without_datadog(*args)
```

Comment:
@p-lambert ouch, I think it should be outside the `ensure` otherwise the return value is discarded.... my bad :(